src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -354,11 +356,14 @@ std::shared_ptr<rpc::GcsNodeInfo> GcsNodeManager::RemoveNode(
 
 void GcsNodeManager::OnNodeFailure(const NodeID &node_id,
                                    const StatusCallback &node_table_updated_callback) {
-  rpc::NodeDeathInfo death_info = InferDeathInfo(node_id);
-  SetDeathInfo(node_id, death_info);
-  bool is_expected_termination =
-      (death_info.reason() == rpc::NodeDeathInfo::AUTOSCALER_DRAIN_PREEMPTED);
-  if (auto node = RemoveNode(node_id, is_expected_termination)) {
+  auto maybe_node = GetAliveNode(node_id);
+  if (maybe_node.has_value()) {
+    rpc::NodeDeathInfo death_info = InferDeathInfo(node_id);
+    auto node = RemoveNode(node_id, death_info);
+    if (!node) {
+      return;
+    }
```

Comment:
this must be non-None?

Reply:
I think you are right, since the GCS event loop is single-threaded, there should be no race here.